pingcap/tidb
# *: fix snapshot problem in show table status

### What problem does this PR solve? 
```
1、drop table t1;
2. SET SESSION tidb_snapshot = 'xxx'; (set snapshot before drop)
3. SHOW TABLE STATUS; (no table t1)
```


### What is changed and how it works?
There are three bugs:
1. When execute sql for `SHOW TABLE STATUS`, uses `ExecRestrictedSQL` instead of `ExecRestrictedSQLWithSnapshot`. 
2. In `ExecRestrictedSQLWithSnapshot`, we don't get the `SnapshotInfoschema`. 
3. When `getRows` for infoschemaTable, we always use the txn's InfoSchema.

Besides, move function `GetInfoSchema` into package `infoSchema" to avoid circular reference


### Check List 

Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)

Note that this pr will revert pr [7459](https://github.com/pingcap/tidb/pull/7459/files)， since pr7459 doesn't have test, I manually test and pass [it](https://github.com/pingcap/tidb/issues/7421).

Code changes

 - Has exported function/method change


Side effects


Related changes

 - Need to cherry-pick to the release branch

Release note

 - Fix snapshot problem in show table status
